### PR TITLE
[OGR] reset layername after SELECT filter is removed

### LINKED ISSUE #56345: Duplicate Layer does not work on layers loaded with SQL Queries from GPKG
### What is the bug or the crash?

If I load a layer with SQL Query (from the Browser panel) starting from a GeoPackage and then duplicate it, and in the duplicated layer I delete the query, a different layer is displayed.

GPKG for example
[limitiISTAT2023.zip](https://github.com/qgis/QGIS/files/14277854/limitiISTAT2023.zip)

https://github.com/qgis/QGIS/assets/7631137/32d2f61c-1d55-4464-9dd1-bda717333230

### Steps to reproduce the issue

1. Start QGIS
2. from the layers panel, create a SQL query in a GeoPakage and load the layer;
3. duplicate the layer you just loaded in the layers panel and delete the query
4. the layer does not display the geometry of the starting layer, but will display a random layer present in the geopackage

## OSGeo4W 64b - win 11

QGIS version
3.35.0-Master
QGIS code revision
26ff0b7235
Qt version
5.15.3
Python version
3.9.18
GDAL/OGR version
3.9.0dev-d7aed6e0b0
PROJ version
9.3.1
EPSG Registry database version
v10.098 (2023-11-24)
GEOS version
3.12.1-CAPI-1.18.1
SQLite version
3.41.1
PDAL version
2.6.0
PostgreSQL client version
16.1
SpatiaLite version
5.1.0
QWT version
6.1.6
QScintilla2 version
2.13.4
OS version
Windows 10 Version 2009

This copy of QGIS writes debugging output.

Active Python plugins
hsa
0.1
db_manager
0.1.20
grassprovider
2.12.99
MetaSearch
0.3.6
processing
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

the problem is also present in QGIS 3.34.3

### PR DESCRIPTION
**Pull request by elpaso**
Fix #56345

When a provider subset string like "SELECT* FROM table WHERE ...." was removed, no `layername=xxxx` was added to the URI, making it ambiguous in case of multiple tables in a dataset (e.g. GPKG).

### CODE REVIEW

File: src/core/vector/qgsvectorlayer.cpp [hunk 1]
```diff
@@ -1266,6 +1266,7 @@ bool QgsVectorLayer::setSubsetString( const QString &subset )
 
   if ( res )
   {
+    mWkbType = mDataProvider->wkbType();
```
Comment (rouault) [inline src/core/vector/qgsvectorlayer.cpp:1269]: that change seems unrelated

Comment (rouault) [reply src/core/vector/qgsvectorlayer.cpp:1269]: or maybe it is. please ignore

**Comment by nirvn** [comment]
@elpaso , not a bad solution at all :)

**Review by rouault — APPROVED** [review]

File: src/core/vector/qgsvectorlayer.cpp (hunk 1 above)
Comment (nyalldawson) [inline src/core/vector/qgsvectorlayer.cpp:1269]: I think there's a little more required here to support changing geometry types via queries -- eg we'd need to rebuild the layer renderer from a default if the geometry type changes. Could this change be split off into a separate PR?

Comment (elpaso) [reply src/core/vector/qgsvectorlayer.cpp]: done

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
@elpaso looks like a valid test failure here.

**Comment by elpaso** [comment]
> @elpaso looks like a valid test failure here.

I cannot reproduce the test failure locally.

**Comment by agiudiceandrea** [comment]
@elpaso why don't just add the `layername=` part when the layer is added via Browser or DB manager? Should'n it be easier to implement?

**Comment by elpaso** [comment]
> @elpaso why don't just add the `layername=` part when the layer is added via Browser or DB manager? Should'n it be easier to implement?

That would not fix the issue because in case of a "SELECT..." filter a different table could be selected and in that case we want to set the correct "layername"  when the filter is removed.

This patch fixes the issue by parsing the query filter and getting the table from there. It will work in most cases, it won't in case of multiple tables in the FROM clause.

Anyway, I cannot reproduce the CI test error locally and I don't know when I'll find the time to work on this issue again.

**Comment by agiudiceandrea** [comment]
My suggestion is to be consistent with the the already present behaviour:
- add a GPKG sub layer to the project -> source: `/path/file_name.gpkg|layername=sub_layer_name`
- add any SQL filter to the layer (even selecting from another sub layer) -> source: `/path/file_name.gpkg|layername=sub_layer_name|subset=the_SQL_query`
- remove the filter from the layer -> source: `/path/file_name.gpkg|layername=sub_layer_name`

Otherwise I think also the already present behaviour should be changed.
